tpdatasrc/tpgamefiles/scr/spell_utils.py

```diff
@@ -0,0 +1,275 @@
+    evt_obj.data1 = spell_id
+    return 0
+
+# if a charge attack is added, and it's not the spell _this_
+# condition was added from, the spell needs to end.
+def touchTouchAttackAdded(attachee, args, evt_obj):
+    spell_id = args.get_arg(0)
+
+    if evt_obj.data1 != spell_id:
+        args.condition_remove()
+        args.remove_spell()
+    return 0
+
+# On the round a charge is held, a free touch attack may be
+# made, since it is supposed to happen as part of casting.
+def touchAdd(attachee, args, evt_obj):
+    spell_id = args.get_arg(0)
+
+    tbFlags = tpdp.cur_seq_get_turn_based_status_flags()
+    tbFlags = tbFlags | TBSF_TouchAttack
+    tpdp.cur_seq_set_turn_based_status_flags(tbFlags)
+    attachee.d20_send_signal(S_TouchAttackAdded, spell_id)
+    return 0
+
+# End the spell and remove the holding-charge condition.
+def End(attachee, args, evt_obj):
+    args.condition_remove()
+    args.remove_spell()
+    return 0
+
+# Common code for handling touch attack charges.
+#
+# 1) Decrements number of charges if positive
+# 2) Ends the condition if charges fall to 0
+# 3) Checks spell resistance on the target.
+# 4) If resistance fails, adds the target to the
+#        spell's target list
+#
+# Returns whether the spell was resisted so that
+# further effects can be run.
+#
+# Common use would be something like:
+#
+#  def OnTouch(attachee, args, evt_obj):
+#    resisted = touchPre(args, evt_obj)
+#
+#    if not resisted:
+#      <spell effects>
+def touchPre(args, evt_obj):
+    action = evt_obj.get_d20_action()
+    caster = action.performer
+    target = action.target
+
+    spell_id = args.get_arg(0)
+    duration = args.get_arg(1)
+    charges = args.get_arg(2)
+
+    packet = tpdp.SpellPacket(spell_id)
+
+    if charges > 0:
+        args.set_arg(2, charges-1)
+
+    resisted = packet.check_spell_resistance(target)
+
+    if charges == 1:
+        packet.end_target_particles(caster)
+        packet.remove_target(caster)
+        args.condition_remove()
+
+    if resisted:
+        game.particles('Fizzle', target)
+        game.sound(7461,1)
+    else:
+        packet.add_target(target, 0)
+
+    packet.update_registry()
+    if packet.target_count <= 0:
+        args.spell_remove()
+
+    return resisted
+
+# Modifiers representing a 'held charge' touch spell
+class TouchModifier(PythonModifier):
+    # Touch modifiers have at least 3 arguments
+    #
+    #  0: spell_id
+    #  1: duration, -1 to not display
+    #  2: num_charges, -1 for until duration expires
+    #
+    # the given argument number is for additional arguments beyond
+    # this.
+    #
+    # The standard countdown hook is not added, since the typical
+    # mechanic is for charges to be held indefinitely until used.
+    # An argument is reserved for time limited touch spells,
+    # however, and the countdown hook can be added in such a case.
+    def __init__(self, name, argn):
+        PythonModifier.__init__(self, name, 3+argn)
```

Comment:
Something to keep in mind: by default this adds an 'Add Only Once' hook. What would happen if you cast the same spell twice?
@dolio

Reply:
This was intentional (if you're talking about what I think you're talking about).

Casting a spell while you're holding charge causes you to lose the held charge. So there should never be a situation where you have two copies of the 'holding charge' condition for a spell.

Reply:
I think we're talking about something different.

PythonModifier has an optional arg preventDup, which is true by default. This adds a generic hook for ET_OnConditionAddPre, which prevents the new condition from being added.

My main concern is that when the 2nd spell instance is cast it may produce an unterminated spell, since the spell cast goes through as normal but the condition isn't actually applied and so no duration countdown actually occurs.

The "preventDup" hook fires before the touchAdd (ET_OnConditionAdd) is called so the S_TouchAttackAdded signal will not fire, and consequently neither will touchTouchAttackAdded which removes the old charge.

In general, I think most spells do not have the preventDup generic hook, so spell modifiers should have this off by default as well.
Incidentally, I just checked the spell condition list which ones have it (0x100ECF30), and it looks like Call Lightning and Call Lightning Storm has this issue as well.

Reply:
Okay, so should spell modifiers in general not use the generic hook, because they need a more specific hook that e.g. cancels the spell in addition to not adding the condition?

And in this case, the older spell should cancel itself via the other hooks I've added, so the new spell doesn't need to cancel itself or prevent its effect from being added?

Reply:
So, I just tried out my spell that uses this, and it seems like it actually works correctly. When I cast the spell a second time, the original spell stops ticking, and the new one starts ticking (based on the `OnBeginRound` functions in the spell scripts).

I'm not sure how that's happening exactly. It seems to work the same when I tell it not to add the generic hook.

Reply:
Oh, it's through the S_Spell_Cast signal which ends the previous spell whenever any spell is cast. This is invoked during the PerformCastSpell action so that's why it happens beforehand.
Although that brings up a different issue, I think this will also happen when any spell is cast on you as well (as a target, rather than caster, that is).

Reply:
Yeah, you're right. Casting a spell on the person holding the charge eliminates the charge. I think Touch of Fatigue has this same problem.

Is there a way to just detect when the subject of the condition casts a spell? Is it just a matter of inspecting the d20 action in the event object to see who is performing the spell casting?

Reply:
Heh, apparently it's like that for most vanilla spells. Another one for the list :)

You should be able to compare attachee == spell_packet.caster in the S_Spell_Cast handler.
@doug1234 might want to fix that in Touch of Fatigue too.

Reply:
I'll take a look at that after I finish up that draft pull request.